azure/notebooks/azure_mnmg.ipynb

```diff
@@ -0,0 +1,674 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "cu_rf_params = {\n",
+    "    'n_estimators': 100,\n",
+    "    'max_depth': 16,\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The cell below creates a client with the cluster we defined earlier in the notebook. Note that we have `cluster.scale`. This is the step where the workers are allocated.\n",
+    "\n",
+    "Once workers become available, we can now run the rest of our workflow - reading and cleaning the data, splitting into training and validation sets, fitting a RF model and predicting on the validation set. We print out the MSE metric for this problem. Note that for better performance we should perform HPO ideally. \n",
+    "\n",
+    "\n",
+    "Refer to the notebooks in the repository for how to perform automated HPO [using RayTune](https://github.com/rapidsai/cloud-ml-examples/blob/main/ray/notebooks/Ray_RAPIDS_HPO.ipynb) and [using Optuna](https://github.com/rapidsai/cloud-ml-examples/blob/main/optuna/notebooks/optuna_rapids.ipynb)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "RMSE:  3.3357165\n"
+     ]
+    }
+   ],
+   "source": [
+    "with Client(cluster) as client:\n",
+    "    import dask_cudf\n",
+    "    cluster.scale(2)\n",
+    "    client.wait_for_workers(2)\n",
+    "    from cuml.dask.ensemble import RandomForestRegressor\n",
+    "\n",
+    "    base_path = 'gcs://anaconda-public-data/nyc-taxi/csv/'\n",
+    "    df_2014 = dask_cudf.read_csv(base_path+'2014/yellow_tripdata_2014*.csv', n_rows=100000)\n",
+    "\n",
+    "    df_2014 = clean(df_2014, remap, must_haves)\n",
+    "\n",
+    "    taxi_df = df_2014.map_partitions(add_features)\n",
+    "\n",
+    "    taxi_df = taxi_df.dropna()\n",
+    "    taxi_df = taxi_df.astype(\"float32\")\n",
+    "    X, y = taxi_df.drop([\"fare_amount\"], axis=1), taxi_df[\"fare_amount\"].astype('float32')\n",
+    "\n",
+    "    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, shuffle=True)\n",
+    "\n",
+    "    workers = client.has_what().keys()\n",
+    "\n",
+    "    X_train, X_test, y_train, y_test = dask_utils.persist_across_workers(client,\n",
+    "                                                           [X_train, X_test, y_train, y_test],\n",
+    "                                                           workers=workers)\n",
+    "    cu_dask_rf = RandomForestRegressor(**cu_rf_params, ignore_empty_partitions=True)\n",
+    "    cu_dask_rf = cu_dask_rf.fit(X_train, y_train)\n",
+    "\n",
+    "    y_pred = cu_dask_rf.predict(X_test)\n",
+    "    \n",
+    "    _y_pred, _y_test = y_pred.compute().to_array(), y_test.compute().to_array()\n",
+    "    \n",
+    "    score = mean_squared_error(_y_pred, _y_test)\n",
+    "    print(\"RMSE: \", np.sqrt(score))"
```

Comment:
Getting the following error after creating VMs:

> Creating worker instance
Network interface ready
Creating VM
Created VM dask-166bf743-worker-8d4bda86
Creating worker instance
Network interface ready
Creating VM
Created VM dask-166bf743-worker-e45ec8b3

KeyError                                  Traceback (most recent call last)
<ipython-input-12-229a9f8e9f62> in <module>
     24                                                            workers=workers)
     25     cu_dask_rf = RandomForestRegressor(**cu_rf_params, ignore_empty_partitions=True)
---> 26     cu_dask_rf = cu_dask_rf.fit(X_train, y_train)
     27 
     28     y_pred = cu_dask_rf.predict(X_test)

/opt/conda/envs/rapids/lib/python3.8/site-packages/cuml/dask/ensemble/randomforestregressor.py in fit(self, X, y, convert_dtype)
    261         """
    262         self.internal_model = None
--> 263         self._fit(model=self.rfs,
    264                   dataset=(X, y),
    265                   convert_dtype=convert_dtype)

/opt/conda/envs/rapids/lib/python3.8/site-packages/cuml/dask/ensemble/base.py in _fit(self, model, dataset, convert_dtype)
    116                 self.client.submit(
    117                     _func_fit,
--> 118                     model[worker],
    119                     worker_data,
    120                     convert_dtype,

KeyError: 'tcp://10.1.0.9:35773'

Reply:
That's probably an error on Azure side with how to workers are setup. Re-running the cell should fix it. Let me know if it persists. I haven't been able to reproduce it reliably.

Reply:
Modifying `client.wait_for_workers(2)` with number of GPU workers (4) resolved this issue, still need to verify we can successfully train the model